src/mlpack/methods/ne/genome.hpp

```diff
@@ -0,0 +1,329 @@
+      if (aNeuronGenes[i].Type() == OUTPUT)
+        ++numOutput;
+    }
+    return numOutput;
+  }
+
+  // Set output length.
+  void NumOutput(ssize_t numOutput) { aNumOutput = numOutput; }
+
+  // Set fitness.
+  void Fitness(double fitness) { aFitness = fitness; }
+
+  // Get fitness.
+  double Fitness() const { return aFitness; }
+
+  // Set adjusted fitness.
+  void AdjustedFitness(double adjustedFitness) { aAdjustedFitness = adjustedFitness; }
+
+  // Get adjusted fitness.
+  double AdjustedFitness() const { return aAdjustedFitness; }
+
+  // Get neuron number.
+  ssize_t NumNeuron() const {
+    return aNeuronGenes.size();
+  }
+  
+  // Get link number.
+  ssize_t NumLink() const {
+    return aLinkGenes.size();
+  }
+
+  // Whether specified neuron id exist in this genome.
+  bool HasNeuronId(ssize_t id) const {
+    for (ssize_t i=0; i<NumNeuron(); ++i) {
+      if (aNeuronGenes[i].Id() == id) {
+        return true;
+      }
+    }
+    
+    return false;
+  }
+
+  // Get neuron by id.
+  NeuronGene GetNeuronById(ssize_t id) {
+    assert(HasNeuronId(id));
+
+    for (ssize_t i=0; i<NumNeuron(); ++i) {
+      if (aNeuronGenes[i].Id() == id) {
+        return aNeuronGenes[i];
+      }
+    }
+  }
+
+  // Get neuron index by id.
+  ssize_t GetNeuronIndex(ssize_t id) const {
+    for(ssize_t i=0; i < NumNeuron(); ++i) {
+        if (aNeuronGenes[i].Id() == id) {
+            return i;
+        }
+    }
+
+    return -1;  // Id start from 0.
+  }
+
+  // Get link index by innovation id.
+  ssize_t GetLinkIndex(ssize_t innovId) const {
+    for(ssize_t i=0; i < NumLink(); ++i) {
+        if (aLinkGenes[i].InnovationId() == innovId) {
+            return i;
+        }
+    }
+
+    return -1;  // Id start from 0.
+  }
+
+  // Whether link exist and enabled.
+  bool ContainEnabledLink(ssize_t innovId) const {
+    for(ssize_t i=0; i < NumLink(); ++i) {
+        if (aLinkGenes[i].InnovationId() == innovId &&
+            aLinkGenes[i].Enabled()) {
+            return true;
+        }
+    }
+    return false;
+  }
+
+  // Whether link exist.
+  bool ContainLink(ssize_t innovId) const {
+    for(ssize_t i=0; i < NumLink(); ++i) {
+        if (aLinkGenes[i].InnovationId() == innovId) {
+            return true;
+        }
+    }
+    return false;
+  }
+
+  // Set neurons' input and output to zero.
+  void Flush() {
+    for (ssize_t i=0; i<aNeuronGenes.size(); ++i) {
+      aNeuronGenes[i].Activation(0);
+      aNeuronGenes[i].Input(0);
+    }
+  }
+
+  // Sort link genes by toNeuron's depth.
+  void SortLinkGenes() {
+    struct DepthAndLink
+    {
+      double depth;
+      LinkGene link;
+ 
+      DepthAndLink(double d, LinkGene& l) : depth(d), link(l) {}
+
+      bool operator < (const DepthAndLink& dL) const
+      {
+        return (depth < dL.depth);
+      }
+    };
+
+    std::vector<double> toNeuronDepths;
+    for (ssize_t i=0; i<aLinkGenes.size(); ++i) {
+      NeuronGene toNeuron = GetNeuronById(aLinkGenes[i].ToNeuronId());
```

Comment:
Maybe it's a good idea, to add a GetDepthById function, in this case we could avoid to copy the object.